src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            try
+            {
+                RuntimeHelpers.EnsureSufficientExecutionStack();
+            }
+            catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public void RunOnEmptyStack<T1, T2>(Action<T1, T2> action, T1 arg1, T2 arg2)
+        {
+            RunOnEmptyStackCore(s =>
+            {
+                var t = (Tuple<Action<T1, T2>, T1, T2>)s;
+                t.Item1(t.Item2, t.Item3);
+                return default(object);
```

Comment:
I can either reuse the same `RunOnEmptyStackCore` method that uses a `Task<R>` or duplicate the code in there to have an overload for a non-generic `Task`. I assume the only difference is that the former will use a tiny bit more memory for the `m_result` field. Any strong opinions on the trade-off between assembly code size and runtime memory usage?